internal/lib/checkpoint.go

```diff
@@ -42,6 +42,35 @@ func (c *ContainerServer) ContainerCheckpoint(
 		return "", fmt.Errorf("container %s is not running", ctr.ID())
 	}
 
+	// At this point the container needs to be paused. As we first checkpoint
+	// the processes in the container and the container will continue to run
+	// after checkpointing, there is a chance that the changed files we include
+	// in the checkpoint archive might change by the now again running processes
+	// in the container.
+	// Assuming this uses runc/crun (the OCI runtime is currently the only one
+	// supporting checkpointing), PauseContainer() will use the cgroup freezer
+	// to freeze the processes. CRIU will also use the cgroup freezer to freeze
+	// the processes if possible. If the cgroup is already frozen by runc/crun
+	// CRIU will not change the freezer status.
+	if err = c.runtime.PauseContainer(ctx, ctr); err != nil {
+		return "", fmt.Errorf("failed to pause container %q before checkpointing: %w", ctr.ID(), err)
+	}
+	defer func() {
+		if err := c.runtime.UpdateContainerStatus(ctx, ctr); err != nil {
+			log.Errorf(ctx, "Failed to update container status: %q: %v", ctr.ID(), err)
+		}
```

Comment:
same here, is this one necessary?

Reply:
Yes, this one is needed. For some reason the CRI-O internal state of the checkpointed container is "running". Only after `UpdateContainerStatus()` it has the correct value of "paused". Not sure why the state is wrong. Especially as during actual checkpointing (`runc checkpoint`) the state is only changed to "stopped" if `KeepRunning` is false.

It is always false in CRI-O (in contrast to Podman) as that is the only mode supported by the KEP and the CRI.